sw/device/lib/crypto/impl/rsa.c

```diff
@@ -263,6 +423,200 @@ otcrypto_status_t otcrypto_rsa_private_key_from_exponents(
+  HARDENED_TRAP();
+  return OTCRYPTO_FATAL_ERR;
+}
+
+otcrypto_status_t otcrypto_rsa_private_key_deconstruct(
+    otcrypto_blinded_key_t *private_key, otcrypto_word32_buf_t p,
+    otcrypto_word32_buf_t q, otcrypto_word32_buf_t d_p,
+    otcrypto_word32_buf_t d_q, otcrypto_word32_buf_t i_q) {
+  if (private_key == NULL || private_key->keyblob == NULL || p.data == NULL ||
+      q.data == NULL || d_p.data == NULL || d_q.data == NULL ||
+      i_q.data == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+  // Entropy complex must be initialized for `hardened_memcpy`.
+  HARDENED_TRY(entropy_complex_check());
+
+  // Infer the RSA size from the public key modulus.
+  otcrypto_rsa_size_t size;
+  HARDENED_TRY(rsa_size_from_private_key(private_key, &size));
+
+  // Ensure that the cofactors, private exponent components, and CRT coefficient
+  // are all the proper length with respect to each other.
+  if (p.len != q.len || d_p.len != p.len || d_q.len != q.len ||
+      i_q.len != p.len) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the caller-provided private key buffer.
+  HARDENED_TRY(private_key_structural_check(size, private_key));
+
+  // Extract the values of p, q, d_p, d_q, and i_q based on the inferred size.
+  switch (size) {
+    case kOtcryptoRsaSize2048: {
+      if (p.len != kRsa2048NumWords / 2) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      rsa_2048_private_key_t *sk =
+          (rsa_2048_private_key_t *)private_key->keyblob;
+      hardened_memcpy(p.data, sk->p.data, p.len);
+      hardened_memcpy(q.data, sk->q.data, q.len);
+      hardened_memcpy(d_p.data, sk->d_p.data, d_p.len);
+      hardened_memcpy(d_q.data, sk->d_q.data, d_q.len);
+      hardened_memcpy(i_q.data, sk->i_q.data, i_q.len);
+      return OTCRYPTO_OK;
+    }
+    case kOtcryptoRsaSize3072: {
+      if (p.len != kRsa3072NumWords / 2) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      rsa_3072_private_key_t *sk =
+          (rsa_3072_private_key_t *)private_key->keyblob;
+      hardened_memcpy(p.data, sk->p.data, p.len);
+      hardened_memcpy(q.data, sk->q.data, q.len);
+      hardened_memcpy(d_p.data, sk->d_p.data, d_p.len);
+      hardened_memcpy(d_q.data, sk->d_q.data, d_q.len);
+      hardened_memcpy(i_q.data, sk->i_q.data, i_q.len);
+      return OTCRYPTO_OK;
+    }
+    case kOtcryptoRsaSize4096: {
+      if (p.len != kRsa4096NumWords / 2) {
+        return OTCRYPTO_BAD_ARGS;
+      }
+      rsa_4096_private_key_t *sk =
+          (rsa_4096_private_key_t *)private_key->keyblob;
+      hardened_memcpy(p.data, sk->p.data, p.len);
+      hardened_memcpy(q.data, sk->q.data, q.len);
+      hardened_memcpy(d_p.data, sk->d_p.data, d_p.len);
+      hardened_memcpy(d_q.data, sk->d_q.data, d_q.len);
+      hardened_memcpy(i_q.data, sk->i_q.data, i_q.len);
+      return OTCRYPTO_OK;
+    }
+    default: {
+      // Invalid key size. Since the size was inferred, should be unreachable.
+      HARDENED_TRAP();
+      return OTCRYPTO_FATAL_ERR;
+    }
```

Comment:
nit: the functions above have both this trap case at the end of the switch _and_ a trap case at the end of the function. Either version is okay with me (it's basically a minor code size vs defensive coding tradeoff) but I think it would be good to stay consistent.